orbit/pkg/update/nudge.go

```diff
@@ -0,0 +1,94 @@
+package update
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/rs/zerolog/log"
+)
+
+// NudgeConfigFetcher is a kind of middleware that wraps an OrbitConfigFetcher and a Runner.
+// It checks the config supplied by the wrapped OrbitConfigFetcher to detects whether the Fleet
+// server has supplied a Nudge config. If so, it sets Nudge as a target on the wrapped Runner.
+type NudgeConfigFetcher struct {
+	// Fetcher is the OrbitConfigFetcher that will be wrapped. It is responsible
+	// for actually returning the orbit configuration or an error.
+	Fetcher OrbitConfigFetcher
+
+	// UpdateRunner is the wrapped Runner where Nudge will be set as a target. It is responsible for
+	// actually ensuring that Nudge is installed and updated via the designated TUF server.
+	UpdateRunner *Runner
+}
+
+func ApplyNudgeConfigFetcherMiddleware(f OrbitConfigFetcher, u *Runner) OrbitConfigFetcher {
+	return &NudgeConfigFetcher{Fetcher: f, UpdateRunner: u}
+}
+
+// GetConfig calls the wrapped Fetcher's GetConfig method, and detects if the
+// Fleet server has supplied a Nudge config. It also handles cases where the Nudge application
+// should be added or removed. If a Nudge config is supplied, it ensures that Nudge is
+// installed and updated via the designated TUF server. If Nudge is already installed but a Nudge
+// config is not supplied, Nudge is removed. If Nudge is removed, the current Orbit process
+// is interrupted.
+func (n *NudgeConfigFetcher) GetConfig() (*fleet.OrbitConfig, error) {
+	cfg, err := n.Fetcher.GetConfig()
+	if err != nil {
+		log.Info().Err(err).Msg("calling GetConfig from NudgeConfigFetcher")
+		return nil, err
+	}
+
+	if cfg == nil {
+		log.Debug().Msg("NudgeConfigFetcher received nil config")
+		return nil, nil
+	}
+
+	var foundTarget bool
+	for _, t := range n.UpdateRunner.opt.Targets {
```

Comment:
This needs to be mutex-protected, otherwise AFAICT a read here could happen at a same time as a write in UpdateRunner.Add/Remove target, which is a data race.

Reply:
agreed, I think this should be moved into `UpdateRunner` itself.